File: examples/chatbots/python/interactive_oauth.py
"""
Interactive OAuth server client for MCP servers requiring OAuth authentication.

This client handles the complete OAuth flow including browser-based authorization.
"""

import logging
import threading
import time
import webbrowser
import httpx
from datetime import timedelta
from http.server import BaseHTTPRequestHandler, HTTPServer
from typing import Optional
from urllib.parse import parse_qs, urlparse

from mcp.client.auth import (
    OAuthClientProvider,
    TokenStorage,
)
from mcp.client.streamable_http import streamablehttp_client, MCP_PROTOCOL_VERSION
from mcp.shared.auth import (
    OAuthClientInformationFull,
    OAuthClientMetadata,
    OAuthToken,
    ProtectedResourceMetadata,
)
from mcp.types import LATEST_PROTOCOL_VERSION


class InMemoryTokenStorage(TokenStorage):
    """
    Simple in-memory token storage implementation.
    In production, you should persist tokens securely. However, for
    the demo chatbot, it's ok to ask the user to re-authenticate
    in the browser each time they run the chatbot.
    """

    def __init__(self):
        self._client_info: Optional[OAuthClientInformationFull] = None
        self._token: Optional[OAuthToken] = None

    async def get_client_info(self) -> Optional[OAuthClientInformationFull]:
        return self._client_info

    async def set_client_info(self, client_info: OAuthClientInformationFull) -> None:
        self._client_info = client_info

    async def get_token(self) -> Optional[OAuthToken]:
        return self._token

    async def set_token(self, token: OAuthToken) -> None:
        self._token = token

    async def clear_token(self) -> None:
        self._token = None


class CallbackHandler(BaseHTTPRequestHandler):
    """HTTP request handler for OAuth callback."""

    def __init__(self, request, client_address, server, callback_data):
        """Initialize with callback data storage."""
        self.callback_data = callback_data
        super().__init__(request, client_address, server)

    def do_GET(self):
        """Handle GET request for OAuth callback."""
        parsed_url = urlparse(self.path)
        query_params = parse_qs(parsed_url.query)

        if parsed_url.path == "/callback":
            # Extract authorization code and state
            auth_code = query_params.get("code", [None])[0]
            state = query_params.get("state", [None])[0]
            error = query_params.get("error", [None])[0]

            if error:
                self.callback_data["error"] = error
                self.send_response(400)
                self.send_header("Content-type", "text/html")
                self.end_headers()
                self.wfile.write(
                    f"<html><body><h1>Authorization Failed</h1><p>Error: {error}</p></body></html>".encode()
                )
            elif auth_code:
                self.callback_data["code"] = auth_code
                self.callback_data["state"] = state
                self.send_response(200)
                self.send_header("Content-type", "text/html")
                self.end_headers()
                self.wfile.write(
                    b"<html><body><h1>Authorization Successful</h1><p>You can close this window.</p></body></html>"
                )
            else:
                self.send_response(400)
                self.send_header("Content-type", "text/html")
                self.end_headers()
                self.wfile.write(
                    b"<html><body><h1>Authorization Failed</h1><p>No authorization code received.</p></body></html>"
                )
        else:
            self.send_response(404)
            self.end_headers()

    def log_message(self, format, *args):
        """Suppress default logging."""
        pass


class CallbackServer:
    """HTTP server for handling OAuth callbacks."""

    def __init__(self, port=8090):
        self.port = port
        self.callback_data = {}
        self.server = None
        self.thread = None

    def start(self):
        """Start the callback server."""
        handler = lambda *args: CallbackHandler(*args, self.callback_data)
        self.server = HTTPServer(("localhost", self.port), handler)
        self.thread = threading.Thread(target=self.server.serve_forever)
        self.thread.daemon = True
        self.thread.start()
        logging.debug(f"Callback server started on port {self.port}")

    def stop(self):
        """Stop the callback server."""
        if self.server:
            self.server.shutdown()
            self.server.server_close()
        if self.thread:
            self.thread.join(timeout=1)
        logging.debug("Callback server stopped")

    def wait_for_callback(self, timeout=300):
        """Wait for OAuth callback with timeout."""
        start_time = time.time()
        while time.time() - start_time < timeout:
            if "code" in self.callback_data:
                return self.callback_data["code"]
            elif "error" in self.callback_data:
                raise ValueError(f"OAuth error: {self.callback_data['error']}")
            time.sleep(0.1)
        raise TimeoutError("OAuth callback timeout")

    def get_state(self):
        """Get the state parameter from callback."""
        return self.callback_data.get("state")


class InteractiveOAuthClient:
    """
    Manages OAuth authentication for MCP servers requiring interactive OAuth.

    This client handles the complete OAuth flow including browser-based authorization.
    """

    def __init__(self, name: str, server_url: str, client_id: str = None):
        self.name = name
        self.server_url = server_url
        self.client_id = client_id
        self.callback_port = 8090
        self.callback_url = f"http://localhost:{self.callback_port}/callback"

    async def create_transport(self):
        """Create OAuth-authenticated transport for MCP communication."""
        if not self.server_url:
            raise ValueError(
                "The server_url must be a valid string and cannot be undefined."
            )

        logging.debug(f"Connecting to OAuth-protected MCP server: {self.server_url}")

        # Create OAuth client metadata
        client_metadata_dict = {
            "client_name": f"MCP Client - {self.name}",
            "redirect_uris": [self.callback_url],
            "grant_types": ["authorization_code", "refresh_token"],
            "response_types": ["code"],
            "token_endpoint_auth_method": "none",  # No client secret for example servers
        }

        # Add client_id to metadata if available
        if self.client_id:
            client_metadata_dict["client_id"] = self.client_id

        # Create callback server
        callback_server = CallbackServer(port=self.callback_port)
        callback_server.start()

        async def callback_handler() -> tuple[str, Optional[str]]:
            """Wait for OAuth callback and return auth code and state."""
            logging.debug("Waiting for authorization callback...")
            try:
                auth_code = callback_server.wait_for_callback(timeout=300)
                return auth_code, callback_server.get_state()
            finally:
                callback_server.stop()

        async def redirect_handler(authorization_url: str) -> None:
            """Handle OAuth redirect by opening browser."""
            logging.debug(f"OAuth redirect handler called - opening browser")
            self._open_browser(authorization_url)

        # Create OAuth authentication handler
        storage = InMemoryTokenStorage()

        # If we have a client_id, create and store client info to skip client registration
        if self.client_id:
            client_info = OAuthClientInformationFull(
                client_id=self.client_id,
                client_id_issued_at=int(time.time()),
                client_name=client_metadata_dict["client_name"],
                redirect_uris=client_metadata_dict["redirect_uris"],
                grant_types=client_metadata_dict["grant_types"],
                response_types=client_metadata_dict["response_types"],
                token_endpoint_auth_method=client_metadata_dict[
                    "token_endpoint_auth_method"
                ],
            )
            # Store client info synchronously before creating the OAuth provider
            await storage.set_client_info(client_info)
            logging.debug(f"Pre-configured client info with ID: {self.client_id}")

        oauth_auth = OAuthClientProvider(
            server_url=self.server_url,
            client_metadata=OAuthClientMetadata.model_validate(client_metadata_dict),
            storage=storage,
            redirect_handler=redirect_handler,
            callback_handler=callback_handler,
        )

        # Discover the required scope from the server and re-create the provider with the scope
        scope = await self.discover_scope(self.server_url, oauth_auth)
        logging.debug(f"Discovered scope from server metadata: {scope}")

        client_metadata_dict["scope"] = scope
        if self.client_id:
            client_info.scope = scope

        oauth_auth = OAuthClientProvider(
            server_url=self.server_url,
            client_metadata=OAuthClientMetadata.model_validate(client_metadata_dict),
            storage=storage,
            redirect_handler=redirect_handler,
            callback_handler=callback_handler,
        )

        logging.debug("Creating transport with OAuth provider...")
        return streamablehttp_client(
            url=self.server_url,
            auth=oauth_auth,
            timeout=timedelta(seconds=60),
        )

    async def discover_scope(
        self, server_url: str, oauthProvider: OAuthClientProvider
    ) -> str:
        """Discovers the required scope from OAuth protected resource metadata."""
        logging.debug("Discovering OAuth metadata...")

        async with httpx.AsyncClient() as client:
            headers = {MCP_PROTOCOL_VERSION: LATEST_PROTOCOL_VERSION}
            response = await client.get(
                server_url, headers=headers, follow_redirects=True
            )

            discovery_request = await oauthProvider._discover_protected_resource(
                response
            )
            logging.debug(
                f"Discovery request: {discovery_request.method} {discovery_request.url}"
            )
            discovery_response = await client.send(discovery_request)
            if discovery_response.status_code != 200:
                logging.warning(
                    f"Response code from discovery request was {discovery_response.status_code}. Using empty scope."
                )
                return ""

            content = await discovery_response.aread()
            resource_metadata = ProtectedResourceMetadata.model_validate_json(content)

            if not resource_metadata.scopes_supported:
                logging.warning(
                    "No scopes found in OAuth protected resource metadata. Using empty scope."
                )
                return ""

            discovered_scope = " ".join(resource_metadata.scopes_supported)
            logging.debug(f"Discovered scope: {discovered_scope}")
            return discovered_scope

    def _open_browser(self, url: str) -> None:
        """Open the authorization URL in the user's default browser."""
        logging.debug(f"Opening browser for authorization: {url}")
        try:
            webbrowser.open(url)
        except Exception as error:
            logging.error(f"Failed to open browser: {error}")
            logging.info(f"Please manually open: {url}")

File: examples/chatbots/typescript/src/interactive_oauth.ts
import { createServer } from "node:http";
import { URL } from "node:url";
import { exec } from "node:child_process";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";
import {
  OAuthClientInformation,
  OAuthClientInformationFull,
  OAuthClientMetadata,
  OAuthTokens,
} from "@modelcontextprotocol/sdk/shared/auth.js";
import {
  OAuthClientProvider,
  UnauthorizedError,
  discoverOAuthProtectedResourceMetadata,
  extractResourceMetadataUrl,
} from "@modelcontextprotocol/sdk/client/auth.js";
import logger from "./logger.js";

export class InteractiveOAuthClient extends StreamableHTTPClientTransport {
  name: string;
  serverUrl: string;
  clientId?: string;
  private oauthProvider?: InMemoryOAuthClientProvider;
  private callbackPort: number;
  private callbackUrl: string;
  private initialized = false;

  constructor(name: string, serverUrl: string, clientId?: string) {
    // Don't call super yet - we'll initialize after OAuth flow
    super(new URL(serverUrl), {});
    this.name = name;
    this.serverUrl = serverUrl;
    this.clientId = clientId;

    this.callbackPort = 8090;
    this.callbackUrl = `http://localhost:${this.callbackPort}/callback`;
  }

  async initialize(): Promise<void> {
    if (this.initialized) return;

    try {
      logger.debug(`Connecting to OAuth-protected MCP server: ${this.serverUrl}`);

      const scope = await this.discoverScope();

      const clientMetadata: OAuthClientMetadata = {
        client_name: `MCP Client - ${this.name}`,
        redirect_uris: [this.callbackUrl],
        grant_types: ["authorization_code", "refresh_token"],
        response_types: ["code"],
        token_endpoint_auth_method: "none",
        scope: scope,
      };

      logger.debug("Creating OAuth provider...");
      this.oauthProvider = new InMemoryOAuthClientProvider(
        this.callbackUrl,
        clientMetadata,
        this.clientId,
        undefined,
        (redirectUrl: URL) => {
          logger.debug(`OAuth redirect handler called - opening browser`);
          this.openBrowser(redirectUrl.toString());
        }
      );

      logger.debug("Starting OAuth flow...");
      await this.performOAuthFlow();
      this.initialized = true;
    } catch (error) {
      logger.error(`Error initializing OAuth server ${this.name}: ${error}`);
      throw error;
    }
  }

  /**
   * Opens the authorization URL in the user's default browser
   */
  private async openBrowser(url: string): Promise<void> {
    logger.debug(`Opening browser for authorization: ${url}`);

    const command = `open "${url}"`;

    exec(command, (error) => {
      if (error) {
        logger.error(`Failed to open browser: ${error.message}`);
        logger.info(`Please manually open: ${url}`);
      }
    });
  }

  /**
   * Starts a temporary HTTP server to receive the OAuth callback
   */
  private async waitForOAuthCallback(): Promise<string> {
    return new Promise<string>((resolve, reject) => {
      const server = createServer((req, res) => {
        // Ignore favicon requests
        if (req.url === "/favicon.ico") {
          res.writeHead(404);
          res.end();
          return;
        }

        logger.debug(`Received OAuth callback: ${req.url}`);
        const parsedUrl = new URL(req.url || "", "http://localhost");
        const code = parsedUrl.searchParams.get("code");
        const error = parsedUrl.searchParams.get("error");

        if (code) {
          logger.debug(
            `Authorization code received: ${code?.substring(0, 10)}...`
          );
          res.writeHead(200, { "Content-Type": "text/html" });
          res.end(`
            <html>
              <body>
                <h1>Authorization Successful!</h1>
                <p>You can close this window and return to the application.</p>
                <script>setTimeout(() => window.close(), 2000);</script>
              </body>
            </html>
          `);

          server.close(() => {
            resolve(code);
          });
        } else if (error) {
          logger.error(`Authorization error: ${error}`);
          res.writeHead(400, { "Content-Type": "text/html" });
          res.end(`
            <html>
              <body>
                <h1>Authorization Failed</h1>
                <p>Error: ${error}</p>
              </body>
            </html>
          `);
          server.close(() => {
            reject(new Error(`OAuth authorization failed: ${error}`));
          });
        } else {
          logger.error(`No authorization code or error in callback`);
          res.writeHead(400);
          res.end("Bad request");
          server.close(() => {
            reject(new Error("No authorization code provided"));
          });
        }
      });

      server.listen(this.callbackPort, () => {
        logger.debug(
          `OAuth callback server started on http://localhost:${this.callbackPort}`
        );
      });
    });
  }

  /**
   * Discovers the required scope from OAuth protected resource metadata
   */
  private async discoverScope(): Promise<string> {
    logger.debug("Making initial request to discover OAuth metadata...");

    const response = await fetch(this.serverUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ jsonrpc: "2.0", method: "ping", id: 1 }),
    });
    const resourceMetadataUrl = extractResourceMetadataUrl(response);

    logger.debug(`Discovered resource metadata URL: ${resourceMetadataUrl}`);
    logger.debug("Fetching OAuth protected resource metadata...");

    const resourceMetadata = await discoverOAuthProtectedResourceMetadata(
      this.serverUrl,
      { resourceMetadataUrl }
    );

    if (
      !resourceMetadata.scopes_supported ||
      resourceMetadata.scopes_supported.length === 0
    ) {
      logger.warn(
        "No scopes found in OAuth protected resource metadata. Using empty scope."
      );
      return "";
    }

    const discoveredScope = resourceMetadata.scopes_supported.join(" ");
    logger.debug(`Discovered scope from server metadata: ${discoveredScope}`);
    return discoveredScope;
  }

  /**
   * Performs OAuth flow without connecting the transport
   */
  private async performOAuthFlow(): Promise<void> {
    logger.debug("Creating temporary transport for OAuth flow...");
    const baseUrl = new URL(this.serverUrl);
    const transport = new StreamableHTTPClientTransport(baseUrl, {
      authProvider: this.oauthProvider,
    });

    const tempClient = new Client(
      { name: "typescript-chatbot", version: "0.1.0" },
      { capabilities: { sampling: {} } }
    );

    try {
      logger.debug("Attempting connection to trigger OAuth flow...");
      await tempClient.connect(transport);
      logger.debug("OAuth flow completed successfully");
      
      // Now reinitialize this transport with the authenticated provider
      Object.assign(this, new StreamableHTTPClientTransport(baseUrl, {
        authProvider: this.oauthProvider,
      }));
    } catch (error) {
      if (error instanceof UnauthorizedError) {
        logger.debug("OAuth required - waiting for authorization...");
        const callbackPromise = this.waitForOAuthCallback();
        const authCode = await callbackPromise;
        await transport.finishAuth(authCode);
        logger.debug("Authorization completed successfully");
        logger.debug("Retrying OAuth flow...");
        await this.performOAuthFlow();
      } else {
        logger.error("OAuth flow failed:", error);
        throw error;
      }
    }
  }
}

/**
 * In-memory OAuth client provider for interactive OAuth flows.
 * In production, you should persist tokens securely. However, for
 * the demo chatbot, it's ok to ask the user to re-authenticate
 * in the browser each time they run the demo.
 */
class InMemoryOAuthClientProvider implements OAuthClientProvider {
  private _clientInformation?: OAuthClientInformationFull;
  private _tokens?: OAuthTokens;
  private _codeVerifier?: string;

  constructor(
    private readonly _redirectUrl: string | URL,
    private readonly _clientMetadata: OAuthClientMetadata,
    private readonly _clientId?: string,
    private readonly _clientSecret?: string,
    onRedirect?: (url: URL) => void
  ) {
    this._onRedirect =
      onRedirect ||
      ((url) => {
        console.log(`Redirect to: ${url.toString()}`);
      });

    // Fill in OAuth client ID and client secret if provided
    if (this._clientId) {
      this._clientInformation = {
        client_id: this._clientId,
        client_secret: this._clientSecret,
        ...this._clientMetadata,
      };
    }
  }

  private _onRedirect: (url: URL) => void;

  get redirectUrl(): string | URL {
    return this._redirectUrl;
  }

  get clientMetadata(): OAuthClientMetadata {
    return this._clientMetadata;
  }

  clientInformation(): OAuthClientInformation | undefined {
    return this._clientInformation;
  }

  saveClientInformation(clientInformation: OAuthClientInformationFull): void {
    this._clientInformation = clientInformation;
  }

  tokens(): OAuthTokens | undefined {
    return this._tokens;
  }

  saveTokens(tokens: OAuthTokens): void {
    this._tokens = tokens;
  }

  redirectToAuthorization(authorizationUrl: URL): void {
    this._onRedirect(authorizationUrl);
  }

  saveCodeVerifier(codeVerifier: string): void {
    this._codeVerifier = codeVerifier;
  }

  codeVerifier(): string {
    if (!this._codeVerifier) {
      throw new Error("No code verifier saved");
    }
    return this._codeVerifier;
  }
}
